_studio/mfx_lib/shared/src/libmfxsw_encode.cpp

```diff
@@ -169,57 +294,17 @@ mfxStatus MFXVideoENCODE_Query(mfxSession session, mfxVideoParam *in, mfxVideoPa
         else
 #endif
         {
-            switch (out->mfx.CodecId)
-            {
-#ifdef MFX_ENABLE_H264_VIDEO_ENCODE
-            case MFX_CODEC_AVC:
-#if defined (MFX_ENABLE_H264_VIDEO_ENCODE_HW)
-                if (!session->m_pENCODE.get())
-                    mfxRes = MFXHWVideoENCODEH264::Query(session->m_pCORE.get(), in, out);
-                else
-                    mfxRes = MFXHWVideoENCODEH264::Query(session->m_pCORE.get(), in, out, session->m_pENCODE.get());
-#else //MFX_VA
-                mfxRes = MFXVideoENCODEH264::Query(in, out);
-#endif //MFX_VA
-                break;
-#endif
-
-#ifdef MFX_ENABLE_MPEG2_VIDEO_ENCODE
-            case MFX_CODEC_MPEG2:
-                mfxRes = MFXVideoENCODEMPEG2_HW::Query(session->m_pCORE.get(), in, out);
-                break;
-#endif
-
-#if defined(MFX_ENABLE_MJPEG_VIDEO_ENCODE)
-            case MFX_CODEC_JPEG:
-                mfxRes = MFXVideoENCODEMJPEG_HW::Query(session->m_pCORE.get(), in, out);
-                break;
-#endif // MFX_ENABLE_MJPEG_VIDEO_ENCODE
-
-#if defined(MFX_ENABLE_H265_VIDEO_ENCODE)
-            case MFX_CODEC_HEVC:
-            {
 #if defined (MFX_ENABLE_HEVC_VIDEO_FEI_ENCODE)
-                bool * feiEnabled = (bool*)session->m_pCORE->QueryCoreInterface(MFXIFEIEnabled_GUID);//required to check FEI plugin registration.
-                MFX_CHECK_NULL_PTR1(feiEnabled);
-                if(*feiEnabled)
-                    mfxRes = MfxHwH265FeiEncode::H265FeiEncode_HW::Query(session->m_pCORE.get(), in, out);
-                else
+            // required to check FEI plugin registration
+            bool *feiEnabled = (bool*)session->m_pCORE->QueryCoreInterface(MFXIFEIEnabled_GUID);
+            MFX_CHECK_NULL_PTR1(feiEnabled);
+            const bool fei = *feiEnabled;
+#else
+            const bool fei = false;
 #endif
-                    mfxRes = MfxHwH265Encode::MFXVideoENCODEH265_HW::Query(session->m_pCORE.get(), in, out);
-                break;
-            }
-#endif // MFX_ENABLE_H265_VIDEO_ENCODE
-
-#if defined(MFX_ENABLE_VP9_VIDEO_ENCODE)
-            case MFX_CODEC_VP9:
-                mfxRes = MfxHwVP9Encode::MFXVideoENCODEVP9_HW::Query(&session->m_coreInt, in, out);
-                break;
-#endif //MFX_ENABLE_VP9_VIDEO_ENCODE
-
-            default:
-                mfxRes = MFX_ERR_UNSUPPORTED;
-            } // switch (out->mfx.CodecId)
+            auto handlers = codecId2Handlers.find(CodecKey(out->mfx.CodecId, fei));
+            mfxRes = handlers == codecId2Handlers.end() ? MFX_ERR_UNSUPPORTED
+                : (handlers->second.primary.query)(session, in, out);
```

Comment:
`auto handlers` is wrong name, it should be `auto handle` - yous search **single** handle.

@mgonchar : should it be `auto& handle`?

Reply:
I disagree: in each `Handlers` structure we have bunch of handlers.

Reply:
You search for the single entry in the codecIdHandlers table. As a result handlers is misleading to me.

Reply:
@dvrogozh why reference, it is an iterator with cheap copy. Moreover as function doesn't return reference the legal is only assignment to const reference